internal/cache/file/downloader/job_test.go

```diff
@@ -405,3 +407,328 @@ func (jt *jobTest) Test_downloadObjectAsync_ErrorWhenFileCacheHasLessSize() {
+	time.Sleep(time.Second * 2)
+	// verify file
+	jt.verifyFile(objectContent)
+	// verify file info cache
+	jt.verifyFileInfoEntry(uint64(objectSize))
+}
+
+func (jt *jobTest) Test_Download_WhenAlreadyDownloading() {
+	// Create new object in bucket and create new job for it.
+	objectName := "path/in/gcs/foo.txt"
+	objectSize := 50 * MiB
+	objectContent := bytes.Repeat([]byte("t"), objectSize)
+	jt.initJobTest(objectName, objectContent, 100, uint64(2*objectSize))
+	// start download but not wait for download
+	ctx := context.Background()
+	jobStatus, err := jt.job.Download(ctx, 1, false)
+	ExpectEq(nil, err)
+	ExpectEq(DOWNLOADING, jobStatus.Name)
+
+	// Again call download but wait for download this time.
+	offset := int64(25 * MiB)
+	jobStatus, err = jt.job.Download(ctx, offset, true)
+
+	ExpectEq(nil, err)
+	// verify that jobStatus is downloading and downloaded at least 25 Mib.
+	ExpectEq(DOWNLOADING, jobStatus.Name)
+	ExpectEq(nil, jobStatus.Err)
+	ExpectGe(jobStatus.Offset, offset)
+	// verify that after some time the whole object is downloaded
+	time.Sleep(time.Second * 2)
+	// verify file
+	jt.verifyFile(objectContent)
+	// verify file info cache
+	jt.verifyFileInfoEntry(uint64(objectSize))
+}
+
+func (jt *jobTest) Test_Download_WhenAlreadyCompleted() {
+	objectName := "path/in/gcs/foo.txt"
+	objectSize := 25 * MiB
+	objectContent := bytes.Repeat([]byte("t"), objectSize)
+	jt.initJobTest(objectName, objectContent, 100, uint64(2*objectSize))
+	// Wait for whole download to be completed.
+	ctx := context.Background()
+	jobStatus, err := jt.job.Download(ctx, int64(objectSize), true)
+	ExpectEq(nil, err)
+	// verify that jobStatus is DOWNLOADING but offset is object size
+	expectedJobStatus := JobStatus{DOWNLOADING, nil, int64(objectSize)}
+	ExpectTrue(reflect.DeepEqual(expectedJobStatus, jobStatus))
+
+	// try to request for some offset when job was already completed.
+	offset := int64(18 * MiB)
+	jobStatus, err = jt.job.Download(ctx, offset, true)
+
+	ExpectEq(nil, err)
+	// verify that jobStatus is completed & offset returned is still 25 MiB
+	// this ensures that async job is not started again.
+	ExpectEq(COMPLETED, jobStatus.Name)
+	ExpectEq(nil, jobStatus.Err)
+	ExpectGe(jobStatus.Offset, objectSize)
+	// verify file
+	jt.verifyFile(objectContent)
+	// verify file info cache
+	jt.verifyFileInfoEntry(uint64(jobStatus.Offset))
+}
+
+func (jt *jobTest) Test_Download_WhenAsyncFails() {
+	objectName := "path/in/gcs/foo.txt"
+	objectSize := 25 * MiB
+	objectContent := bytes.Repeat([]byte("t"), objectSize)
+	// set size of cache smaller than object size.
+	jt.initJobTest(objectName, objectContent, 100, uint64(objectSize-1))
+
+	// Wait for whole download to be completed/failed.
+	ctx := context.Background()
+	jobStatus, err := jt.job.Download(ctx, int64(objectSize), true)
+
+	ExpectEq(nil, err)
+	// verify that jobStatus is failed
+	ExpectEq(FAILED, jobStatus.Name)
+	ExpectEq(ReadChunkSize, jobStatus.Offset)
```

Comment:
Should validate this? It can be anything. 

What's the definition of offset when jobStatus is FAILED?